### PR TITLE
CLDC-3404 Allow support to update user organisation

### PR DESCRIPTION
**Pull request by kosiakkatrina**
Allow support users to change user organisations
image
image
Validate that relevant relationships are there to move the logs over
image
Choosing to unassign logs assigns them to a user in the current user organisation called "Unassigned", some of the organisations might already have this user as it would get created during migration from the old service in cases where no relevant user was created to own that log. If  "Unassigned" user doesn't exist for organisation, we'd create one
image

This PR also allows editing deactivated users by support user as we want to be able to move deactivated users between organisations. This does also mean that other user fields can be edited, I'm yet to check if that's okay, or whether we should do some work to split that up.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -56,20 +56,24 @@ def dpo; end
...
       if @user == current_user
         bypass_sign_in @user
         flash[:notice] = I18n.t("devise.passwords.updated") if user_params.key?("password")
-        if user_params.key?("email")
+        if user_params.key?("email") && user_params[:email] != @user.email
           flash[:notice] = I18n.t("devise.email.updated", email: @user.unconfirmed_email)
         end
 
-        redirect_to account_path
+        if updating_organisation?
```
Comment (RachaelBooth) [inline app/controllers/users_controller.rb:72]: Can a user update their own organisation? I guess yes if they are support.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -144,6 +153,58 @@ def delete
     redirect_to users_organisation_path(@user.organisation), notice: I18n.t("notification.user_deleted", name: @user.name)
   end
 
+  def log_reassignment
+    authorize @user
```
Comment (RachaelBooth) [inline app/controllers/users_controller.rb:157]: Is this not covered by one of the `before_action` things on the controller?

File: app/models/user.rb [hunk 3]
```diff
@@ -270,6 +280,47 @@ def phone_with_extension
...
+      when "reassign_stock_owner"
+        reassign_stock_owners(new_organisation, lettings_logs_to_reassign, sales_logs_to_reassign)
+      when "reassign_managing_agent"
+        reassign_managing_agents(new_organisation, lettings_logs_to_reassign, sales_logs_to_reassign)
+      when "unassign"
+        unassign_organisations(lettings_logs_to_reassign, sales_logs_to_reassign, current_organisation)
+      end
+
+      send_organisation_change_email(current_organisation, new_organisation, log_reassignment, logs_count)
+    rescue ActiveRecord::RecordInvalid => e
```
Comment (RachaelBooth) [inline app/models/user.rb]: Are there any other plausible exceptions that we might need to handle?

File: app/policies/user_policy.rb [hunk 4]
```diff
@@ -10,17 +10,15 @@ def edit_password?
...
-  end
-
   %w[
     edit_roles?
     edit_dpo?
     edit_key_contact?
   ].each do |method_name|
     define_method method_name do
-      (@current_user.data_coordinator? || @current_user.support?) && @user.active?
+      return true if @current_user.support?
```
Comment (RachaelBooth) [inline app/policies/user_policy.rb:19]: Does active not matter for support? (Actually, does it matter for anyone since you can't log in while not active?)

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (kosiakkatrina) [reply app/controllers/users_controller.rb:72]: Hmm, I don't think it was explicit whether they should or shouldn't be able to move themselves in the ACs, but it might be useful on staging if people can move themselves around as we had to manually move people for testing purposes in the past. Let me know if there are any concerns with this tho

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (kosiakkatrina) [reply app/controllers/users_controller.rb:157]: We do run `authenticate_scope!` but for some user paths we use policies instead. I feel like it'a a bit cleaner than adding another if into `authenticate_scope!`

File: app/models/user.rb (hunk 3 above)
Comment (kosiakkatrina) [reply app/models/user.rb]: I changed it to StandardError, I think it should be most errors (?) would that work?

File: app/policies/user_policy.rb (hunk 4 above)
Comment (kosiakkatrina) [reply app/policies/user_policy.rb:19]: Oh, so I don't think it matters for current user, cause as you say, they should be able to login before they even get to this point. But for the user that is getting edited we have an AC to allow deactivated users to be moved. So I'm allowing other fields for deactivated user to be editable too (cause otherwise we'd only show organisation question on the edit page, which is a bit odd that support can change the org but nothing else)

So basically support users can edit deactivated and active users, whereas coordinators can only edit active as it was before.

Hope this makes sense

**Review by RachaelBooth — APPROVED** [review]

File: app/models/user.rb [hunk 5]
```diff
@@ -292,8 +292,8 @@ def reassign_logs_and_update_organisation(new_organisation, log_reassignment)
     return unless new_organisation
 
     ActiveRecord::Base.transaction do
-      lettings_logs_to_reassign = assigned_to_lettings_logs
-      sales_logs_to_reassign = assigned_to_sales_logs
+      lettings_logs_to_reassign = assigned_to_lettings_logs.visible
+      sales_logs_to_reassign = assigned_to_sales_logs.visible
```
Comment (RachaelBooth) [inline app/models/user.rb:296]: Does this cause problems if a user has pending logs from a bulk upload that they may later choose to load?

Comment (kosiakkatrina) [reply app/models/user.rb:296]: Yes, that's a very good point! I'll follow up what we want to do with those on the ticket

File: app/models/forms/bulk_upload_lettings_resume/fix_choice.rb [hunk 6]
```diff
@@ -56,7 +56,7 @@ def save!
       end
 
       def preflight_valid?
-        bulk_upload.choice != "create-fix-inline" && bulk_upload.choice != "bulk-confirm-soft-validations"
+        bulk_upload.choice.blank?
```
Comment (kosiakkatrina) [inline app/models/forms/bulk_upload_lettings_resume/fix_choice.rb:59]: This should basically prevent users from accessing the fix choice page if they made any kind of choice (either manually or by user being moved). I think it makes sense to discredit as many bulk uploads so that they don't linger with a possibility of the user stumbling upon the link and accidentally creating them. 

We might also separately want to update routing here if a log has setup errors, cause I don't think anything is stopping us from navigating to fix-choice page directly (?)
On top of that any bulk uploads for closed collection years should probably also not be accessible, or at least not uploadable
